ecpy_pulses/sequences/base_sequences.py

```diff
@@ -0,0 +1,760 @@
+
+    def _recompute_indexes(self, first_index=0, free_index=None):
+        """ Recompute the item indexes and update the vars of the root_seq.
+
+        Parameters
+        ----------
+        first_index : int, optional
+            Index in items of the first item whose index needs to be updated.
+
+        free_index : int, optional
+            Value of the first free index.
+
+        """
+        if free_index is None:
+            free_index = self.index + 1
+
+        # Cleanup the linkable_vars for all the pulses which will be reindexed.
+        linked_vars = self.root.linkable_vars
+        for var in linked_vars[:]:
+            if var[0].isdigit() and int(var[0]) >= free_index:
+                linked_vars.remove(var)
+
+        for item in self.items[first_index:]:
+
+            item.index = free_index
+            prefix = '{}_'.format(free_index)
+            linkable_vars = [prefix + var for var in item.linkable_vars]
+            linked_vars.extend(linkable_vars)
+
+            if isinstance(item, BaseSequence):
+                item.unobserve('_last_index', self._item_last_index_updated)
+                item._recompute_indexes()
+                item.observe('_last_index', self._item_last_index_updated)
+                free_index = item._last_index + 1
+
+            # We have a non indexed item (pulse or template).
+            else:
+                free_index += 1
+
+        self._last_index = free_index - 1
+
+    def _item_last_index_updated(self, change):
+        """ Update the items indexes whenever the last index of a child
+        sequence is updated.
+
+        """
+        index = self.items.index(change['object']) + 1
+        free_index = change['value'] + 1
+        self._recompute_indexes(index, free_index)
+
+
+class RootSequence(BaseSequence):
+    """ Base of any pulse sequences.
+
+    This Item perform the first step of compilation by evaluating all the
+    entries and then unravelling the pulse sequence (elimination of condition
+    and loop flattening).
+
+    Notes
+    -----
+
+    The linkable_vars of the RootSequence stores all the known linkable vars
+    for the sequence.
+    The start, stop, duration and def_1, def_2 members are not used by the
+    RootSequence. The time_constrained member only affects the use of the
+    sequence duration.
+
+    """
+    # --- Public API ----------------------------------------------------------
+
+    #: Dictionary of external variables whose values should be given before
+    #: the start of the compilation stage.
+    external_vars = Dict(Unicode()).tag(pref=True)
+
+    #: Duration of the sequence when it is fixed. The unit of this time is
+    # fixed by the context.
+    sequence_duration = Unicode().tag(pref=True)
+
+    #: Reference to the executioner context of the sequence.
+    context = Instance(BaseContext)
+
+    index = set_default(0)
+    name = set_default('Root')
+
+    def __init__(self, **kwargs):
```

Comment:
Use a default method rather than overriding the \_\_init\_\_ and set has_root to True by default.